include/mplib/core/attached_body.h

```diff
@@ -0,0 +1,104 @@
+#pragma once
+
+#include <string>
+#include <vector>
+
+#include "mplib/core/articulated_model.h"
+#include "mplib/kinematics/types.h"
+#include "mplib/macros/class_forward.h"
+#include "mplib/types.h"
+#include "mplib/utils/conversion.h"
+
+namespace mplib {
+
+// AttachedBodyTplPtr
+MPLIB_CLASS_TEMPLATE_FORWARD(AttachedBodyTpl);
+
+/**
+ * Object defining bodies that can be attached to robot links.
+ * This is useful when handling objects picked up by the robot.
+ *
+ * Mimicking MoveIt2's ``moveit::core::AttachedBody``
+ *
+ * https://moveit.picknik.ai/main/api/html/classmoveit_1_1core_1_1AttachedBody.html
+ */
+template <typename S>
+class AttachedBodyTpl {
+ public:
+  // Common type alias
+  using CollisionObjectPtr = fcl::CollisionObjectPtr<S>;
+  using ArticulatedModelPtr = ArticulatedModelTplPtr<S>;
+
+  // TODO(merge): Support multi collision geometry
+  /**
+   * Construct an attached body for a specified link.
+   *
+   * @param name: name of the attached body
+   * @param object: collision object of the attached body
+   * @param attached_articulation: robot articulated model to attach to
+   * @param attached_link_id: id of the articulation link to attach to
+   * @param pose: attached pose (relative pose from attached link to object)
+   * @param touch_links: the link names that the attached body touches
+   */
+  AttachedBodyTpl(const std::string &name, const CollisionObjectPtr &object,
+                  const ArticulatedModelPtr &attached_articulation,
+                  int attached_link_id, const Isometry3<S> &pose,
+                  const std::vector<std::string> &touch_links = {});
+
+  /// @brief Gets the attached object name
+  const std::string &getName() const { return name_; }
+
+  /// @brief Gets the attached object (CollisionObjectPtr)
+  CollisionObjectPtr getObject() const { return object_; }
+
+  /// @brief Gets the articulation this body attached to
+  ArticulatedModelPtr getAttachedArticulation() const { return attached_articulation_; }
```

Comment:
I understand this may be a bit unclear. However, moveit also has similar method names: `AttachedBody::getAttachedLinkName()` and `AttachedBody::getAttachedLink()`.

Also, in moveit's setup, it's not possible to attach an articulation to another articulation. Only attaching a vector of shapes to an articulation link is allowed. So there's no confusion there.

Thus, I'll keep this naming for now.

Ref: https://moveit.picknik.ai/main/api/html/classmoveit_1_1core_1_1AttachedBody.html